openverse_catalog/dags/common/sensors.py

```diff
@@ -0,0 +1,93 @@
+import os
+from datetime import timedelta
+from typing import Iterable
+
+from airflow.exceptions import AirflowException
+from airflow.models import DagModel, DagRun
+from airflow.sensors.base import BaseSensorOperator
+from airflow.triggers.temporal import TimeDeltaTrigger
+from airflow.utils.session import provide_session
+from airflow.utils.state import State
+from sqlalchemy import func
+
+
+class ExternalDAGsSensorAsync(BaseSensorOperator):
+    """
+    Waits for a list of different DAGs to not be running, freeing up
+    a worker slot while it's waiting.
+
+    external_dag_ids: A list of dag_ids that you want to wait for
+    :param check_existence: Set to `True` to check if the external task exists (when
+        external_task_id is not None) or check if the DAG to wait for exists (when
+        external_task_id is None), and immediately cease waiting if the external task
+        or DAG does not exist (default value: False).
+    """
+
+    # TODO: Not sure if this is necessary
+    template_fields = ["external_dag_ids"]
+
+    # TODO: I don't understand what operator extra links are, and whether it's
+    # necessary to implement this.
+    # @property
+    # def operator_extra_links(self):
+    #     """Return operator extra links"""
+    #     return [ExternalDAGsSensorLink()]
+
+    def __init__(
+        self,
+        *,
+        external_dag_ids: Iterable[str],
+        check_existence: bool = False,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.external_dag_ids = external_dag_ids
+        self.check_existence = check_existence
+        self._has_checked_existence = False
+
+    @provide_session
+    def execute(self, context, event=None, session=None):
+        self.log.info(
+            "Poking for DAGS %s ... ",
+            self.external_dag_ids,
+        )
+
+        # Check DAG existence only once, on the first execution.
+        if self.check_existence and not self._has_checked_existence:
+            self._check_for_existence(session=session)
+
+        count_running = (
+            session.query(func.count())
+            .filter(
+                DagRun.dag_id.in_(self.external_dag_ids),
+                DagRun.state == State.RUNNING,
+            )
+            .scalar()
+        )
+
+        self.log.info("%s DAGs are in the running state", count_running)
+        if count_running == 0:
+            return True
+
+        # If there are running DAGs, trigger deferral of the Sensor so that
+        # the worker slot will be freed and the pool does not become
+        # deadlocked waiting on this task.
+        self.defer(
+            trigger=TimeDeltaTrigger(
+                timedelta(seconds=5)
```

Comment:
I think we'll want this configurable at the operator level!

Reply:
Now that we're using `reschedule` mode, I've changed this to just use the `poke_interval`!

I think this is what you mean but just confirming -- I didn't think it was necessary to be configurable _per media type_ since it seems like a single interval for the data refresh DAGs would be fine.

I'm keeping it at 5 seconds for now. The thought is that since this is running inside a pool with one slot, at worst the waiting `wait_for_data_refresh` tasks can each run in between the three steps of whatever DAG is actually running (ie they can't interrupt `wait_for_completion` once it starts). Having a short poke interval means that when the `wait_for_completion` finally passes, you won't need to wait a long time for the next data refresh to kick off.

Does that logic make sense to you? 🤔

Reply:
It does! I agree that we won't need this different per media type, but since this was essentially filling in the role of `poke_interval`, I think it would have made sense to pull up as an operator level variable (similar to `external_dag_ids`) so it could be different were we to use this sensor in other contexts.

I will say, though, that I think we can actually have it be 5 minutes rather than 5 seconds. With the data refreshes taking on the order of days (rather than minutes), I don't think we need that fine a resolution once `wait_for_completion` is done. Additionally, rescheduling the task every 5 seconds puts a lot of unnecessary pressure on the Airflow scheduler (see #309 for some other context). So I think we could have a `poke_interval` of 5 minutes at least, if not something like 10 or 15 even.

Reply:
Ah, I was under the impression that because of the singular pool, the task wouldn't continually reschedule while waiting -- but I think I'm confusing that with something I read about the triggers in the original implementation with the Deferred Sensor. That would be a lot of overhead. 10-15 min seems totally reasonable given the length of the refreshes.

This does make testing more annoying, but we can always edit it locally while testing. Maybe an opportunity for a testing env variable?

Reply:
I think a testing env var makes sense here! I would suggest an Airflow Variable as well but we're not likely to update that value in dev more than once I'd imagine.